test/e2e/externalplugin/generate_test.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externalplugin
+
+import (
+	"path/filepath"
+
+	pluginutil "sigs.k8s.io/kubebuilder/v3/pkg/plugin/util"
+
+	//nolint:golint
+	//nolint:revive
+	. "github.com/onsi/ginkgo/v2"
+
+	//nolint:golint
+	//nolint:revive
+	. "github.com/onsi/gomega"
+
+	//nolint:golint
+	//nolint:revive
+	//nolint:golint
+	//nolint:revive
+	"sigs.k8s.io/kubebuilder/v3/test/e2e/utils"
+)
+
+var _ = Describe("kubebuilder", func() {
+	Context("plugin sampleexternalplugin/v1", func() {
+		var (
+			kbc *utils.TestContext
+		)
+
+		BeforeEach(func() {
+			var err error
+			kbc, err = utils.NewTestContext(pluginutil.KubebuilderBinName, "GO111MODULE=on")
+			Expect(err).NotTo(HaveOccurred(), "Prepare NewTestContext should return no error.")
+			Expect(kbc.Prepare()).To(Succeed())
+		})
+
+		AfterEach(func() {
+			kbc.Destroy()
+		})
+
+		It("should generate a runnable project with sample external plugin", func() {
+			GenerateProject(kbc)
+		})
+
+	})
+})
+
+// GenerateProject implements a sampleexternalplugin/v1 external plugin project defined by a TestContext.
+func GenerateProject(kbc *utils.TestContext) {
+	var err error
+
+	By("initializing a project")
```

Comment:
See that before init you need to ensure that the plugin was built. 
It is not better build the plugin here instead of create the sh?
Could we not move the code to here and to do that in go so that is easier to keep maintained as well? 

See that you can run it locally either if you move all to here by just trigging the tests from your IDE

Reply:
I don't disagree that it might be easier to maintain, but I think there is a lot of precedence to build stuff in a shell script for the e2e tests. IMO we shouldn't try to reinvent the wheel by adding this to the go code and follow the existing pattern for building and configuring e2e tests with shell scripts.

Reply:
Hi @everettraven,

> but I think there is a lot of precedence to build stuff in a shell script for the e2e tests.
We do not have any shell that does things which are required for the e2e tests.
You can trigger locally the e2e suite and you will see that you just need to have any k8s up and running to so.

We have shell for generate the samples and a few of them to setup the prow, which does not seems to few in the same scenario.

If @Eileen-Yu make it working using the shell (besides I do not like that because will broke the possibility to run the e2e tests in dev as described above) I am OK with we get it merged and change after in a follow up. However, see that it is not working and seems exactly because those steps are not performed on the e2e test. So, when the init command is called it fails, see:

```
     s: "kubebuilder init --plugins sampleexternalplugin/v1 --domain sample.domain.com failed with error: (exit status 1) Error: no plugin could be resolved with key \"sampleexternalplugin/v1\"\nUsage:\n  kubebuilder [flags]\n\nExamples:\nThe first step is to initialize your project:\n    kubebuilder init [--plugins=<PLUGIN KEYS> [--project-version=<PROJECT VERSION>]]\n\n<PLUGIN KEYS> is a comma-separated list of plugin keys from the following table\nand <PROJECT VERSION> a supported project version for these plugins.\n\n                             Plugin keys | Supported project versions\n-----------------------------------------+----------------------------\n               base.go.kubebuilder.io/v3 |                          3\n               base.go.kubebuilder.io/v4 |                          3\n        declarative.go.kubebuilder.io/v1 |                       2, 3\n deploy-image.go.kubebuilder.io/v1-alpha |                          3\n                    go.kubebuilder.io/v2 |                       2, 3\n                    go.kubebuilder.io/v3 |                          3\n                    go.kubebuilder.io/v4 |                          3\n         grafana.kubebuilder.io/v1-alpha |                          3\n      kustomize.common.kubebuilder.io/v1 |                          3\n      kustomize.common.kubebuilder.io/v2 |                          3\n\nFor more specific help for the init command of a certain plugins and project version\nconfiguration please run:\n    kubebuilder init --help --plugins=<PLUGIN KEYS> [--project-version=<PROJECT VERSION>]\n\nDefault plugin keys: \"go.kubebuilder.io/v4\"\nDefault project version: \"3\"\n\n\nFlags:\n  -h, --help                     help for kubebuilder\n      --plugins strings          plugin keys to be used for this subcommand execution\n      --project-version string   project version (default \"3\")\n\n2023/05/24 13:03:58 no plugin could be resolved with key \"sampleexternalplugin/v1\"\n",
      }
  occurred
```

Reply:
Hey @camilamacedo86 , I'm trying to follow your idea:

Since `test/e2e/utils` provides context to bring rich prerequisites, including `k8s cluster`, `kubectl`, `kubebuilder`.
That is how e2e tests get performed for internal plugins.

So, it is not necessary/consistent to deal with prerequisites on a different pattern, such as running a new script.

Is that something you mentioned of and external plugin should NOT break?
> We currently do not utilize any shell scripting to accomplish the tasks required for our end-to-end (e2e) tests.

---

Here is my opinion:
We have a well-structured directory under `test/e2e`:
1. those single files such as `local.sh`, `setup.sh` are used to provide the general environment for e2e testing
2. plugin related stuff are put into separate files such as `v4`, `deployimage`

To maintain that way, when we create a new e2e test suite for external-plugin, it is not necessary/recommended to touch those public files.

Then, if building and configuring external-plugin is needed, we can follow what we did to other plugins:
1. At the very beginning in the ginkgo context, use go code to build and configure external-plugin
2. At the end of the ginkgo context, delete the temporary asset for external-plugin